### PR TITLE
`openpyxl`: Use duck typing for workbook sheets

### PR DESCRIPTION
**Pull request by Avasam**
Closes https://github.com/python/typeshed/issues/11706

### CODE REVIEW

File: stubs/openpyxl/openpyxl/workbook/workbook.pyi [hunk 1]
```diff
@@ -17,6 +17,22 @@ from openpyxl.worksheet.worksheet import Worksheet
 _WorkbookWorksheet: TypeAlias = Worksheet | WriteOnlyWorksheet | ReadOnlyWorksheet
 _WorkbookSheet: TypeAlias = _WorkbookWorksheet | Chartsheet
 
+# The type of worksheets in a workbook are the same as the `*Param` aliases above.
+# However, because Worksheet adds a lots of attributes that other _WorkbookChild subclasses don't have
+# (ReadOnlyWorksheet doesn't even inherit form it), this ends up being too disruptive
```
Comment (JelleZijlstra) [inline stubs/openpyxl/openpyxl/workbook/workbook.pyi:22]:
```suggestion
# (ReadOnlyWorksheet doesn't even inherit from it), this ends up being too disruptive
```

File: stubs/openpyxl/openpyxl/workbook/workbook.pyi [hunk 2]
```diff
@@ -17,6 +17,22 @@ from openpyxl.worksheet.worksheet import Worksheet
...
+# to the typical usage of openpyxl where sheets are just Worksheets.
+# Using Any may just loose too much type information and duck-typing
+# from Worksheet works great here, even if the instance type might be wrong.
+@type_check_only
+class _WorksheetLike(  # type: ignore[misc]
+    Worksheet, WriteOnlyWorksheet, ReadOnlyWorksheet, Protocol  # pyright: ignore[reportGeneralTypeIssues]
```
Comment (JelleZijlstra) [inline stubs/openpyxl/openpyxl/workbook/workbook.pyi:28]: A Protocol that inherits from a concrete class isn't meaningful. You can ignore away the error, but there is no guarantee about how type checkers will treat such a type. I don't think we should do this.

Comment (Avasam) [reply stubs/openpyxl/openpyxl/workbook/workbook.pyi:28]: In hindsight, I think that the fact this special class inherits from all possible subtypes solves what I was trying to solve without the need for a protocol.

Is has all the attributes of a Worksheet, and works with `isinstance`, or being passed to a function expecting a specific sheet type.

**Review by JelleZijlstra — COMMENTED** [review]
This is an interesting idea, but it feels like something that could well come back to bite us in the future. Any user who starts digging into the types is likely to end up confused.

I'm not completely opposed but I'd like to hear from other maintainers before we commit to this new technique.

**Review by srittau — APPROVED** [review]
Let's try this. I was bitten by this recently when installing types-openpyxl. Maybe this "poor man's `AnyOf`" (python/typing#566) will help.

**Pull request merged**